### DISCUSSION TITLE
Expose API for creating Databases

### DISCUSSION THREAD

**Discussion by craigmulligan** [supabase/supabase#70] [Feature Requests] [answered]
* **I'm submitting a ...**
[ ] bug report
[x] feature request
[ ] question about the decisions made in the repository
[ ] question about how to use this project

* **Summary**

This ticket is to track API access via an API key as well as exposing a way to programmatically create DBs.

**Comment by kiwicopple** [comment]
Thanks for opening @hobochild. Copying across my email comments for visibility:

This is on our roadmap - we are building and using the API internally right now but it's not ready for public use yet. I will continue to send updates to this issue with the timelines

In case it helps: we are even creating an API for the database itself, so you can spin up one database, then programmatically create schemas/tables/columns/etc

**Comment by martineboh** [comment]
@kiwicopple programmatically creating schemas/tables/columns/ could be a win for `Supabase`...think `Prisma2`

**Comment by kiwicopple** [comment]
I agree @martineboh - I've started this work over on https://github.com/supabase/pg-api

☝️  This repo will fit into our larger Supabase API where you can spin up new databases/projects/orgs etc

ps. @martineboh - i still owe you an email/docs for realtime self-hosting. I haven't forgotten! We have been building the releases to make it easier, and our docs will detail how to use the releases: https://github.com/supabase/realtime/releases. More platforms coming soon (DO, AWS etc)

**Comment by steve-chavez** [comment]
Hey guys. Just chiming in with an idea.

> programmatically creating schemas/tables/columns/

[aquameta's meta extension](https://github.com/aquametalabs/aquameta/tree/master/src/pg-extension/meta) might help for this. It converts DDL to DML. Since you use PostREST, this extension might get you a way for creating schemas/tables/columns through [supabase/postgrest-js](https://github.com/supabase/postgrest-js), for free!

**Reply by isauran** [reply]
Actual url: https://github.com/aquametalabs/aquameta/tree/master/extensions/meta

**Comment by kiwicopple** [comment]
thanks Steve that's kinda genius 💯 

We have hopes to support external databases so the extension might be a blocker, but I'll dig into their implementation and figure out if we can make it work. 

Nice to see you in our repos too!

**Comment by ericjeker** [comment]
Started using Supabase yesterday and I love it so far. ❤️ 
Nothing fancy, I just created a Stackbliz project using just a small Supabase Angular service: https://stackblitz.com/edit/supabase-angular-example

I was looking for a way to create schemas and manage migrations, rollbacking changes, etc... something we can trigger when releasing a new version of an app.

Is that something you guys are also working on?

Do you plan on creating a way to add hooks on some API routes? I saw the events, I guess it would be nice to have INSERTING/INSERTED events, like some before/after middlewares. Maybe supporting transactions?

Thanks for the great work! 👍

**Comment by kiwicopple** [comment]
Hey @ericjeker 👋 Glad you're finding it useful! That's an awesome example application. The first Supabase/Angular implementation I've seen so far actually

> a way to create schemas and manage migrations, rollbacking changes, etc

Yes, we will eventually release a Supabase CLI & git integration which handles all of this. For now though, you will have to do it through an alternative migration tool (I use [dbmate](https://github.com/amacneil/dbmate)). You can find the database connection details in the "settings" of your project.

> Do you plan on creating a way to add hooks on some API routes?
If I've understood this correctly, you can [subscribe](https://supabase.io/docs/library/subscribe) to your database events, which gives you the same effect here. This covers the `INSERTED` usecase, but not the `INSERTED`. Two ways to do this:

- create database triggers in your database (we will eventually add a "trigger editor" to the site)
- create database functions, which are [exposed via the API](https://supabase.io/docs/library/stored-procedures)

**Comment by ericjeker** [comment]
> If I've understood this correctly, you can subscribe to your database events, which gives you the same effect here. This covers the INSERTED usecase, but not the INSERTED. Two ways to do this:

I was thinking about some kind of serverless functions that we could "plug" on the API routes. So they will be on the server-side. A typical use-case for this would be sending an e-mail after a subscription happened (INSERTED) and obviously server-side validation (INSERTING):

-> POST /api/users
1. Trigger INSERTING function or before middleware (authorization, transformation, validation) 
2. INSERT in the DB
3. Trigger INSERTED function or after middleware (send email)

This could also be used for server-side authentication/authorization.
But maybe it's completely out of scope with what you want to create here. I guess I am just throwing some ideas at you. :-)

**Comment by kiwicopple** [comment]
> Trigger INSERTING function or before

This can be done directly in the database using [triggers](https://www.postgresql.org/docs/9.1/sql-createtrigger.html) and we will add a nice UI to make this easy

> Trigger INSERTED function or after

Also can be done with triggers, although we will have a nice interface for this using our Realtime server which you can wire together "flows". Something like zapier:

![image](https://user-images.githubusercontent.com/10214025/87283129-90165300-c527-11ea-9ff8-c89d6300458e.png)

**Reply by miro-ur** [reply]
This looks amazing. Any updates on this ?

**Reply by kiwicopple** [reply]
Still a WIP - we're hoping for Q1/Q2 next year, but we are super busy with the current set of features (auth, database scaling etc).

**Comment by ericjeker** [comment]
OK, well, this looks awesome! Can't wait to test it more. 👍 
In the meantime I will read about triggers and refresh my Postgres skills.

**Comment by tehnorm** [comment]
Has there been any update here? We are looking for a way to provision databases for different environments as part of the development cycle.

**Reply by kiwicopple** [reply]
We're working on a branching workflow now - hopefully will have something for you in Q1 👍

**Reply by knownasilya** [reply]
Awesome! Looking forward to this!

**Reply by royletron** [reply]
hey @kiwicopple I've scoured the blog but didn't see anything about this. Is this is still in progress? It's a bit of a blocker for us adopting Supabase fully (which we really really want to) as we have development across branches - using Vercel - where we need a DB per branch - with the CI provisioning and deleting them.

**Comment by joshnuss** [comment]
Related to creating databases: would be cool if there was a URL-based way for requesting a database.

I'm imagining a similar experience to how DigitalOcean or Heroku setup hosting via a link. It's really nice to have that one-click experience you can put in a readme.

Example flow:

1. User clicks the setup link, ie `http://app.supabase.io/setup?ddl=domain.tld/path/to/file.sql&name=optional-name`
2. User is prompted to login or create account
2. User is sees a grant screen asking for permission. Ideally they're shown the SQL (or list of objects it will create)
3. User click "OK" and the database is created and all the commands from are run

In some ways it's similar to the samples that are shown the first time you create a query, just taking it one step further. It may also relate to the `supa` CLI, maybe it's a layer over that?

**Comment by kcc989** [comment]
Hey @kiwicopple any update on creating DBs via the API? I'm evaluating DB providers right now and really like supabase. However I need to be able to spin up new dbs automatically.

**Reply by nicksrandall** [reply]
I need this feature as well. It would enable supabase to operate as a "SaaSaaS" provider (ie other SaaS companies can provide a postgrest table to their customers via Supabase).

**Reply by MatthewHager** [reply]
We'd love to be able to OAuth against the user's org and then be able to project config like adding tables, managing storage, etc... All the things you can do in the Supabase backend, but programmatically.

**Reply by tonyxiao** [reply]
Second this one as well. Ideally with an option to either create in our own supabase org or customer's supabase org. Perhaps even creating one  in our org to start but transfering to customer org when desired.

**Comment by Pramuspl** [comment]
I need it too. Are there any updates?

**Comment by mfrfinbox** [comment]
Same here! It will be so great to be able to do this.

**Comment by chreds** [comment]
This would be great.

**Comment by inian** [comment, marked as answer]
Projects can be created programmatically via the management API - https://supabase.com/docs/reference/api/create-a-project
